### PR TITLE
Add docker based cluster usage for running tempto-examples tests

### PR DESCRIPTION
**Pull request by kokosing**
Add docker based cluster usage for running tempto-examples tests

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -18,9 +18,6 @@ out/
 # tests logging output
 tempto_logs
 
-# tempto local configuration
```
Comment (sopel39) [inline .gitignore]: isn't it needed?

File: tempto-examples/bin/run_on_docker.sh [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  docker exec $(hadoop_master_container) hive -e 'show tables'
+}
+
+function docker_compose_network() {
+  docker network ls | grep docker | cut  -f 1 -d ' '
+}
+
+function check_presto() {
+  docker run \
+    --rm -it \
```
Comment (sopel39) [inline tempto-examples/bin/run_on_docker.sh]: interactive mode will fail on non-tty terminal. Ask @arhimondr for details

File: tempto-examples/bin/run_on_docker.sh [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    --net $(docker_compose_network) \
+    teradatalabs/presto-cli \
+    java -jar /presto-cli.jar --server presto-master:8080 \
+    --execute "SHOW CATALOGS"  | grep -i hive
+}
+
+function run_product_tests() {
+  docker run \
+    --rm -it \
```
Comment (sopel39) [inline tempto-examples/bin/run_on_docker.sh]: ditto

File: tempto-examples/docker/presto-cli/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,6 @@
+FROM teradatalabs/centos6-java8-oracle
```
Comment (sopel39) [inline tempto-examples/docker/presto-cli/Dockerfile]: where is this dockerfile used?

File: tempto-examples/bin/run_on_docker.sh (hunk 2 above)
Comment (kokosing) [inline tempto-examples/bin/run_on_docker.sh]: Why should I care non-tty terminals? It designed to be run on travis or by developer locally.

File: tempto-examples/bin/run_on_docker.sh [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+function docker_compose_network() {
+  docker network ls | grep docker | cut  -f 1 -d ' '
+}
+
+function check_presto() {
+  docker run \
+    --rm -it \
+    --net $(docker_compose_network) \
+    teradatalabs/presto-cli \
```
Comment (sopel39) [inline tempto-examples/bin/run_on_docker.sh]: can't you just launch command on running `presto-master` container? `presto-cli` image wouldn't be needed then

File: tempto-examples/docker/docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  presto-master:
+    hostname: presto-master
+    build: 'presto-server'
+    ports: 
+      - '8080:8080'
+    depends_on:
+      - 'hadoop-master'
+      - 'psql1'
+      - 'psql2'
+      - 'ssh'
```
Comment (sopel39) [inline tempto-examples/docker/docker-compose.yml]: why `presto-master` depends on `ssh`?

File: tempto-examples/bin/run_on_docker.sh (hunk 2 above)
Comment (sopel39) [inline tempto-examples/bin/run_on_docker.sh]: This is similar to what is in Presto so it would be good to make it work the same. I am not sure you can expect it to run only on Travis and by developer. It works on Travis (I am not sure if by design), but would fail on Jenkins.

File: tempto-examples/bin/run_on_docker.sh (hunk 5 above)
Comment (kokosing) [inline tempto-examples/bin/run_on_docker.sh]: I don't want to mix responsibilities. Also presto-master does not have presto-jdbc jar or any client jars.

File: .gitignore (hunk 1 above)
Comment (losipiuk) [inline .gitignore]: A comment?

File: tempto-examples/bin/run_on_docker.sh [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+      break
+    fi
+    sleep 5
+  done
+
+  return ${EXIT_CODE}
+}
+
+function hadoop_master_container(){
+  docker ps --format '{{.Names}}' | grep hadoop-master
```
Comment (losipiuk) [inline tempto-examples/bin/run_on_docker.sh]: That will be better test:
```
docker ps -f name=hadoop-master -q |grep .
```

File: tempto-examples/bin/run_on_docker.sh [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+function check_hive() {
+  docker exec $(hadoop_master_container) hive -e 'show tables'
+}
+
+function docker_compose_network() {
+  docker network ls | grep docker | cut  -f 1 -d ' '
```
Comment (losipiuk) [inline tempto-examples/bin/run_on_docker.sh]: I do not have such network on mac.
Does it appear only if I run some docker compose?

Also I do not like this `grep`. It is very error prone. 
Can't we use `docker network inspect ...` instead?

File: .travis.yml [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+  - oraclejdk8
+
+sudo: required
+dist: trusty
+
+services:
+  - docker
+
+before_install:
+  # This section can be removed once travis get equipped with docker 1.10 and docker-compose 1.6
```
Comment (losipiuk) [inline .travis.yml]: Is there an older version of docker on travis?
Or are we installing our own.
If the latter why cannot we install newer version?

File: tempto-examples/bin/run_on_docker.sh [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    --net $(docker_compose_network) \
+    teradatalabs/presto-cli \
+    java -jar /presto-cli.jar --server presto-master:8080 \
+    --execute "SHOW CATALOGS"  | grep -i hive
+}
+
+function run_product_tests() {
+  docker run \
+    --rm -it \
+    --net $(docker_compose_network) \
```
Comment (sopel39) [inline tempto-examples/bin/run_on_docker.sh]: why you don't just link connectors together by name instead of using `grep`?

File: build.gradle [hunk 11]
```diff
@@ -43,7 +43,7 @@ ext.versions = [
         commons_lang3: '3.3.2',
         commons_dbcp2: '2.1',
         hive         : '0.13.1-3',
-        presto       : '0.97',
+        presto       : '0.143',
```
Comment (losipiuk) [inline build.gradle]: Unrelated change?

File: tempto-core/src/main/java/com/teradata/tempto/internal/listeners/ProgressLoggingListener.java [hunk 12]
```diff
@@ -87,8 +86,15 @@ public void onTestSkipped(ITestResult testCase)
 
     private void logTestEnd(String outcome)
     {
-        long executionTime = System.currentTimeMillis() - testStartTime;
-        LOGGER.info("{}     /    took {}", outcome, DurationFormatUtils.formatDuration(executionTime, "m' minutes and 's' seconds'"));
+        if (LOGGER.isInfoEnabled()) {
+            long executionTime = System.currentTimeMillis() - testStartTime;
+            if (executionTime > 1000) {
```
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/internal/listeners/ProgressLoggingListener.java]: why so?

File: tempto-examples/bin/run_on_docker.sh [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+# try to stop already running containers
+docker-compose -f "${DOCKER_COMPOSE_LOCATION}" stop || true
+
+docker-compose -f "${DOCKER_COMPOSE_LOCATION}" pull
+docker-compose -f "${DOCKER_COMPOSE_LOCATION}" build
+
+docker-compose -f "${DOCKER_COMPOSE_LOCATION}" up -d 
+docker-compose -f "${DOCKER_COMPOSE_LOCATION}" logs --no-color &
+
+docker build -t teradatalabs/presto-cli "${PRODUCT_TESTS_ROOT}"/docker/presto-cli
```
Comment (losipiuk) [inline tempto-examples/bin/run_on_docker.sh]: don't we need to some cleanup after build here

File: tempto-examples/docker/presto-server/etc/catalog/jmx.properties [hunk 14]
```diff
@@ -0,0 +1 @@
+connector.name=jmx
```
Comment (losipiuk) [inline tempto-examples/docker/presto-server/etc/catalog/jmx.properties]: missing new line

**Comment by losipiuk** [comment]
looks like an improvement :)

File: tempto-examples/bin/run_on_docker.sh (hunk 7 above)
Comment (arhimondr) [inline tempto-examples/bin/run_on_docker.sh]: `docker-compose ps -q hadoop-master`

File: tempto-examples/bin/run_on_docker.sh (hunk 2 above)
Comment (arhimondr) [inline tempto-examples/bin/run_on_docker.sh]: Please check this out https://github.com/Teradata/presto/pull/180

File: tempto-examples/bin/run_on_docker.sh (hunk 8 above)
Comment (arhimondr) [inline tempto-examples/bin/run_on_docker.sh]: You can create yet another service in docker-compose, and run what you want with `docker-compose-run service-name`. So you don't need to determine a network and specify it for docker all around.

File: tempto-examples/bin/run_on_docker.sh (hunk 13 above)
Comment (arhimondr) [inline tempto-examples/bin/run_on_docker.sh]: Shouldn't we add this image to the docker-compose file?

File: tempto-examples/bin/run_on_docker.sh [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+retry check_hive
+retry check_presto
+
+# run product tests
+set +e
+run_product_tests "$*"
```
Comment (arhimondr) [inline tempto-examples/bin/run_on_docker.sh]: This will have problem with handling `INT` or `TERM` signals. You should make a trap.

File: build.gradle (hunk 11 above)
Comment (kokosing) [inline build.gradle]: without this change I would need to use presto-server 0.97. I can extract this.

File: tempto-core/src/main/java/com/teradata/tempto/internal/listeners/ProgressLoggingListener.java (hunk 12 above)
Comment (kokosing) [inline tempto-core/src/main/java/com/teradata/tempto/internal/listeners/ProgressLoggingListener.java]: It is meaningless to display to most of the tests that they last 0 minutes and 0 second.

File: tempto-examples/bin/run_on_docker.sh (hunk 13 above)
Comment (kokosing) [inline tempto-examples/bin/run_on_docker.sh]: do you mean removing docker image? I would not care about it.

Comment (kokosing) [inline tempto-examples/bin/run_on_docker.sh]: do you mean removing docker image? I would not care about it. It is like removing files from local maven repositories.

File: .gitignore (hunk 1 above)
Comment (kokosing) [inline .gitignore]: now tempto-configuration-local.yaml are committed to repo, so ignoring them in git makes it inconvenient

File: .travis.yml (hunk 9 above)
Comment (kokosing) [inline .travis.yml]: There are older versions installed in travis. I would prefer to remove this code and use whatever travis has, hence I stick with oldest possible versions of docker and docker-compose. I hope that waiting period for that version in travis should shorter with that approach.

File: tempto-examples/docker/presto-server/etc/catalog/jmx.properties (hunk 14 above)
Comment (kokosing) [inline tempto-examples/docker/presto-server/etc/catalog/jmx.properties]: done

File: tempto-examples/docker/docker-compose.yml (hunk 6 above)
Comment (kokosing) [inline tempto-examples/docker/docker-compose.yml]: done

File: tempto-examples/bin/run_on_docker.sh (hunk 10 above)
Comment (kokosing) [inline tempto-examples/bin/run_on_docker.sh]: done

File: tempto-examples/bin/run_on_docker.sh (hunk 7 above)
Comment (kokosing) [inline tempto-examples/bin/run_on_docker.sh]: done

File: tempto-examples/bin/run_on_docker.sh (hunk 8 above)
Comment (kokosing) [inline tempto-examples/bin/run_on_docker.sh]: done

File: tempto-examples/bin/run_on_docker.sh (hunk 2 above)
Comment (kokosing) [inline tempto-examples/bin/run_on_docker.sh]: done

File: tempto-examples/bin/run_on_docker.sh (hunk 3 above)
Comment (kokosing) [inline tempto-examples/bin/run_on_docker.sh]: done

File: tempto-examples/bin/run_on_docker.sh (hunk 13 above)
Comment (kokosing) [inline tempto-examples/bin/run_on_docker.sh]: @arhimondr done

File: tempto-examples/bin/run_on_docker.sh (hunk 15 above)
Comment (kokosing) [inline tempto-examples/bin/run_on_docker.sh]: done

File: tempto-core/src/main/java/com/teradata/tempto/internal/listeners/ProgressLoggingListener.java (hunk 12 above)
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/internal/listeners/ProgressLoggingListener.java]: whatever. Just 1s seems very arbitrary to me.

File: tempto-examples/bin/run_on_docker.sh [hunk 16]
```diff
@@ -24,55 +24,47 @@ function retry() {
 }
 
 function hadoop_master_container(){
-  docker ps --format '{{.Names}}' | grep hadoop-master
+  ${DOCKER_COMPOSE} ps -q hadoop-master
```
Comment (losipiuk) [inline tempto-examples/bin/run_on_docker.sh]: does it work?
I tested with `docker` and it always returned status code 0. That is why I suggested adding `| grep .` at the end of command.

**Comment by losipiuk** [comment]
looks good

**Comment by losipiuk** [comment]
though - travis build failed on this PR. This should probably not happen :)

File: tempto-core/src/main/java/com/teradata/tempto/internal/listeners/ProgressLoggingListener.java (hunk 12 above)
Comment (maciejgrzybek) [inline tempto-core/src/main/java/com/teradata/tempto/internal/listeners/ProgressLoggingListener.java]: Can you at least extract this to some constant?

File: tempto-core/src/test/groovy/com/teradata/tempto/assertions/QueryAssertTest.groovy [hunk 17]
```diff
@@ -56,6 +57,11 @@ public class QueryAssertTest
     }
   }
 
+  def setupSpec()
+  {
+    System.setProperty(TEST_CONFIGURATION_URI_KEY, "/configuration/global-configuration-tempto.yaml");
```
Comment (maciejgrzybek) [inline tempto-core/src/test/groovy/com/teradata/tempto/assertions/QueryAssertTest.groovy]: Isn't there any better way than introducing config file to tests?
Cannot we just decouple unit tests from config file?

Comment (kokosing) [inline tempto-core/src/test/groovy/com/teradata/tempto/assertions/QueryAssertTest.groovy]: It is not just simple. QueryAssert is using static methods it is hacky to use mocking there.
